adapters/silvermob/silvermob.go

```diff
@@ -0,0 +1,193 @@
+package silvermob
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type SilverMobAdapter struct {
+	endpoint template.Template
+}
+
+func NewSilverMobBidder(endpointTemplate string) *SilverMobAdapter {
+	template, err := template.New("endpointTemplate").Parse(endpointTemplate)
+	if err != nil {
+		return nil
+	}
+	return &SilverMobAdapter{endpoint: *template}
+}
+
+func GetHeaders(request *openrtb.BidRequest) *http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return &headers
+}
+
+func (a *SilverMobAdapter) MakeRequests(
+	openRTBRequest *openrtb.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo,
+) (
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
+
+	var errors []error
+	var silvermobExt *openrtb_ext.ExtSilverMob
+	var err error
+
+	for i, imp := range openRTBRequest.Imp {
+		silvermobExt, err = a.getImpressionExt(&imp)
```

Comment:
This will result in only the `imp.ext` of the last imp being used, though will report errors if any of the other imps generate them. You might want to ensure that all the zoneid's and hosts match, or potentially split the request into multiple requests, one per unique set of (zoneid, host).

Reply:
Changed this part to processing of only the first element in Imp

Reply:
`MakeRequests()` returns a slice of `RequestData`, so it is easy to just make one request for each imp. PBS will handle merging the bids from the different requests back together. `MakeBids()` will be called with the result of each request that was created, and then merge all the bids made from all the `MakeBids()` calls into the final PBS response. Yours is not the first adapter for a bidder that cannot handle more than one `imp` per request.

Reply:
Thanks, splitted into separate requests